Consistently truncate too wide gate parameters in circuit drawers based on string length

dhruvbhq opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
Truncate too wide gate parameters in all 3 circuit drawers based on length of the resulting string (and not the number of parameters, as was coded previously). Adds the truncation behavior to the text drawer as well, where it was missing.

### Details and comments

Refer to #6364 and #6443. It was discussed that for the matplotlib drawer, drawings for circuits which have too wide parameters should be truncated, as they exceed the width of a single row and cause the circuit to stretch. With this PR, I have tried to address the same issue in the text drawer in case of initialize instruction. In case of the text drawer, when the circuit is initialized with parameters that are too wide, the drawing gets too stretched and eventually occupies multiple lines. Such an excessively stretched circuit drawing impedes efficient visualization for users. As the matplotlib and Latex drawers already have the truncation behavior implemented, this fix truncates the parameters of initialize in the text drawer, in case the length of the resulting string exceeds a particular threshold.

For instance, a sample piece of code which would reproduce the problem (considering a worst case scenario involving complex state initialization) would be:

`import qiskit`
`import numpy as np`
`dim = 3`
`qr = qiskit.QuantumRegister(dim, 'qr')`
`qc = qiskit.QuantumCircuit(qr)`
`init_state = np.zeros(2**dim) + np.zeros(2**dim)*1j`
`for i in range(2**dim):`
`    init_state[i] = (i+1) + (i+1)*1j`
`init_state = init_state / np.linalg.norm(init_state)`
`qc.initialize(init_state)`
`qc.draw(output='text')`

In the existing version, the output is a stretched drawing:
![image](https://user-images.githubusercontent.com/61590679/121820037-1eea9d00-ccae-11eb-84b3-cb74b73edd03.png)


After the fix present in this PR, this gets rendered as:
![image](https://user-images.githubusercontent.com/61590679/121819896-4ab95300-ccad-11eb-9b10-777950d1f42c.png)

I have also added a test case for this scenario.

1ucian0 commented:
Thanks @dhruvbhq  can this be generalized for all the parameters, like in https://github.com/Qiskit/qiskit-terra/pull/6443? From the code, it seems to me that only affects `initialize`.

Cryoris commented:
To add an example to @1ucian0's comment: we mean truncating something like this
```
     ┌──────────────────────────────────────────────────────────────────────────────────────────────────────┐
q_0: ┤ circuit-0(theta[0],theta[1],theta[2],theta[3],theta[4],theta[5],theta[6],theta[7],theta[8],theta[9]) ├
     └──────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

dhruvbhq commented:
@1ucian0 @Cryoris Your suggestion makes sense. I have removed the condition to check for initialize operation. Also added a test case to cover the parametric circuit case shown above.

Let me know what you think.

dhruvbhq commented:
It appears that very recently, the function params_for_label got moved elsewhere (#6078), and merging to remove conflicts in this branch got messed up. Could someone please point me how to clean it up, and where this code got moved to? Thanks :)

boschmitt commented on qiskit/visualization/utils.py:
```diff
@@ -114,10 +114,17 @@ def get_param_str(op, drawer, ndigits=3):
+            truncate_params = (
+                (drawer == "latex" and count > 3)
+                or (drawer == "mpl" and count > 15)
+                or (drawer == "text" and (len(",".join(param_list)) > 30))
```
In every iteration you build the string for the current `param_list`. This code does not need to be particularly fast, but this behaviour seems too wasteful. It would be better to keep an accumulator which you increment based on the length of the parameter you currently add to `param_list`.

boschmitt commented on qiskit/visualization/utils.py:
```diff
@@ -114,10 +114,17 @@ def get_param_str(op, drawer, ndigits=3):
         param_list = [f"{op.params[0]}[{op.unit}]"]
     else:
         param_list = []
+        truncate_params = False
```
Do you really need this variable? (I think its there just to break a long conditional.)

dhruvbhq replied:
If I write the whole conditional inside the `if` statement, I get a pylint error that there are too many boolean expressions in the `if` (max permitted were 5). 

If there is consensus that truncation should happen based on the length of the string, then the logic can be simplified and this variable can be removed.

boschmitt reviewed (commented):
Both `latex` and `mpl` will truncate based on the number of parameters in the list. I find it hard to believe that the problem is the number or parameters and not the length of the resulting string. (Unless `str(param)` already truncate the size of `param`.)

 Wouldn't make sense change the behaviour to be similar to what `text` is doing here?

javabster commented:
Hi @dhruvbhq are you still working on this?

dhruvbhq commented:
Hi @javabster I've lost track of this; please feel free to re-assign.
